### PR TITLE
Feature/nwbv2

### PR DESCRIPTION
**Pull request by ukos-git**
#### Cross links

https://github.com/AllenInstitute/MIES/issues/22
https://github.com/AllenInstitute/MIES/pull/179 (TODO list)
https://github.com/AllenInstitute/IPNWB/issues/5
https://github.com/AllenInstitute/IPNWB/pull/16

Close https://github.com/AllenInstitute/IPNWB/issues/5.

### CODE REVIEW

**Comment by ukos-git** [comment]
rebased on https://github.com/AllenInstitute/IPNWB/tree/make-all-functions-threadsafe

**Comment by ukos-git** [comment]
will there be a separate PR for https://github.com/AllenInstitute/IPNWB/tree/make-all-functions-threadsafe?

**Comment by t-b** [comment]
Rebased.

**Comment by ukos-git** [comment]
I am getting close to spec 2.2.0 but there is an error in the pynwb validator if I add the additional hdmf namespace to the file.

```
▶ python -m pynwb.validate --cached-namespace ~/Documents/Playground/mies-playground/nwbv2/roundtrip/HardwareTests.nwb
Validating /home/matthias/Documents/Playground/mies-playground/nwbv2/roundtrip/HardwareTests.nwb against cached namespace information using namespace hdmf-common.
Traceback (most recent call last):
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/validate/validator.py", line 215, in get_validator
    return self.__validators[dt]
KeyError: 'NWBFile'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/matthias/Documents/Projects/pynwb/src/pynwb/validate.py", line 117, in <module>
    main()
  File "/home/matthias/Documents/Projects/pynwb/src/pynwb/validate.py", line 111, in main
    ret = ret or _validate_helper(io=io, namespace=ns)
  File "/home/matthias/Documents/Projects/pynwb/src/pynwb/validate.py", line 26, in _validate_helper
    errors = validate(**kwargs)
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/utils.py", line 455, in func_call
    return func(**parsed['args'])
  File "/home/matthias/Documents/Projects/pynwb/src/pynwb/__init__.py", line 179, in validate
    return validator.validate(builder)
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/utils.py", line 440, in func_call
    return func(self, **parsed['args'])
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/validate/validator.py", line 233, in validate
    validator = self.get_validator(dt)
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/utils.py", line 440, in func_call
    return func(self, **parsed['args'])
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/validate/validator.py", line 218, in get_validator
    raise ValueError(msg)
ValueError: data type 'NWBFile' not found in namespace hdmf-common
```

pynwb looks for the `NWBFile` in the `hdmf-common` namespace but it is in `nwb.file.yaml` in `core` namespace. The new hdmf namespace is required for 2.2.0 so I think, this is a pynwb problem.

I get the same error on current master on the pynwb example file:

```
~/Documents/Playground/mies-playground/nwbv2    [virtualenv]                                                                          ⍉
▶ python ~/Documents/Projects/pynwb/docs/gallery/domain/icephys.py             

~/Documents/Playground/mies-playground/nwbv2    [virtualenv]                                                                           
▶ python -m pynwb.validate --no-cached-namespace icephys_example.nwb 
Validating icephys_example.nwb against pynwb namespace information using namespace hdmf-common.
Traceback (most recent call last):
  File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/matthias/Documents/Projects/pynwb/src/pynwb/validate.py", line 117, in <module>
    main()
  File "/home/matthias/Documents/Projects/pynwb/src/pynwb/validate.py", line 111, in main
    ret = ret or _validate_helper(io=io, namespace=ns)
  File "/home/matthias/Documents/Projects/pynwb/src/pynwb/validate.py", line 26, in _validate_helper
    errors = validate(**kwargs)
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/utils.py", line 455, in func_call
    return func(**parsed['args'])
  File "/home/matthias/Documents/Projects/pynwb/src/pynwb/__init__.py", line 179, in validate
    return validator.validate(builder)
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/utils.py", line 440, in func_call
    return func(self, **parsed['args'])
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/validate/validator.py", line 232, in validate
    raise ValueError(msg)
ValueError: builder must have data type defined with attribute 'data_type'

~/Documents/Playground/mies-playground/nwbv2    [virtualenv]                                                                          ⍉
▶ python -m pynwb.validate --cached-namespace icephys_example.nwb   
Validating icephys_example.nwb against cached namespace information using namespace hdmf-common.
Traceback (most recent call last):
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/validate/validator.py", line 215, in get_validator
    return self.__validators[dt]
KeyError: 'NWBFile'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/matthias/Documents/Projects/pynwb/src/pynwb/validate.py", line 117, in <module>
    main()
  File "/home/matthias/Documents/Projects/pynwb/src/pynwb/validate.py", line 111, in main
    ret = ret or _validate_helper(io=io, namespace=ns)
  File "/home/matthias/Documents/Projects/pynwb/src/pynwb/validate.py", line 26, in _validate_helper
    errors = validate(**kwargs)
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/utils.py", line 455, in func_call
    return func(**parsed['args'])
  File "/home/matthias/Documents/Projects/pynwb/src/pynwb/__init__.py", line 179, in validate
    return validator.validate(builder)
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/utils.py", line 440, in func_call
    return func(self, **parsed['args'])
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/validate/validator.py", line 233, in validate
    validator = self.get_validator(dt)
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/utils.py", line 440, in func_call
    return func(self, **parsed['args'])
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/validate/validator.py", line 218, in get_validator
    raise ValueError(msg)
ValueError: data type 'NWBFile' not found in namespace hdmf-common
```

**Comment by ukos-git** [comment]
files now validate against cached namespaces.

```
▶ python -m pynwb.validate --cached-namespace HardwareTests.nwb   
Validating HardwareTests.nwb against cached namespace information using namespace core.
 - no errors found.
```

https://pastebin.com/e9Zqewku

File: IPNWB_Utils.ipf [hunk 1]
```diff
@@ -816,11 +816,18 @@ End
...
-	return folder
+	string path
+
+	PathInfo IPNWB_DiscLocation
+	if(V_flag)
+		return s_path
+	endif
+
+	path = GetFolder(FunctionPath(""))
+	NewPath/Q/O IPNWB_DiscLocation, path
```
Comment (ukos-git) [inline IPNWB_Utils.ipf]: You create a new symbolic path but you do not check if the path exists. What is the purpose of the symbolic path in this context?

Comment (t-b) [reply IPNWB_Utils.ipf]: `FunctionPath` in an IM was initially thought of as being buggy. And that symbolic path was meant to cache the result of `FunctionPath`.

But the real bug was that the try/catch statement in the caller of SpecDiscLocation did not clear the RTE so `AbortONRTE` triggered on a lingering RTE.

**Pull request merged**